scc/light_client/provider/rpc_test.go

```diff
@@ -0,0 +1,318 @@
+		provider := NewRpcProviderFromClient(nil)
+		require.True(provider.IsClosed())
+	})
+
+	t.Run("provider with client is not closed", func(t *testing.T) {
+		client := NewMockRpcClient(ctrl)
+		client.EXPECT().Close().AnyTimes()
+		provider := NewRpcProviderFromClient(client)
+		require.False(provider.IsClosed())
+		provider.Close()
+	})
+
+	t.Run("provider with client can be closed", func(t *testing.T) {
+		client := NewMockRpcClient(ctrl)
+		client.EXPECT().Close()
+		provider := NewRpcProviderFromClient(client)
+		require.False(provider.IsClosed())
+		provider.Close()
+		require.True(provider.IsClosed())
+	})
+
+	t.Run("closed provider can be re-closed", func(t *testing.T) {
+		client := NewMockRpcClient(ctrl)
+		client.EXPECT().Close().AnyTimes()
+		provider := NewRpcProviderFromClient(client)
+		provider.Close()
+		require.True(provider.IsClosed())
+		provider.Close()
+		require.True(provider.IsClosed())
+	})
+}
+
+func TestRpcProvider_FailsToRequestAfterClose(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	client := NewMockRpcClient(ctrl)
+
+	provider := NewRpcProviderFromClient(client)
+
+	// close provider
+	client.EXPECT().Close()
+	provider.Close()
+
+	// get committee certificates
+	_, err := provider.GetCommitteeCertificates(0, 1)
+	require.Error(err)
+
+	// get block certificates
+	_, err = provider.GetBlockCertificates(0, 1)
+	require.Error(err)
+}
+
+func TestRpcProvider_NewRpcProvider_ReportsErrorForInvalidURL(t *testing.T) {
+	require := require.New(t)
+
+	provider, err := NewRpcProviderFromURL("not-a-url")
+	require.Error(err)
+	require.Nil(provider)
+}
+
+func TestRpcProvider_GetCertificates_PropagatesErrorFromClientCall(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	client := NewMockRpcClient(ctrl)
+
+	committeeError := fmt.Errorf("committee error")
+	client.EXPECT().Call(gomock.Any(), "sonic_getCommitteeCertificates",
+		gomock.Any(), gomock.Any()).Return(committeeError)
+
+	blockError := fmt.Errorf("block error")
+	client.EXPECT().Call(gomock.Any(), "sonic_getBlockCertificates",
+		gomock.Any(), gomock.Any()).Return(blockError)
+
+	provider := NewRpcProviderFromClient(client)
+
+	// get committee certificates
+	_, err := provider.GetCommitteeCertificates(0, 1)
+	require.ErrorIs(err, committeeError)
+
+	// get block certificates
+	_, err = provider.GetBlockCertificates(0, 1)
+	require.ErrorIs(err, blockError)
+}
+
+func TestRpcProvider_GetCommitteeCertificates_ReportsCorruptedCertificatesOutOfOrder(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	client := NewMockRpcClient(ctrl)
+	provider := NewRpcProviderFromClient(client)
+
+	tests := [][]ethapi.CommitteeCertificate{
+		{
+			makeCommitteeCertForPeriod(1),
+		},
+		{
+			makeCommitteeCertForPeriod(0),
+			makeCommitteeCertForPeriod(2),
+		},
+	}
+
+	for _, committees := range tests {
+
+		// client setup
+		client.EXPECT().Call(gomock.Any(), "sonic_getCommitteeCertificates",
```

Comment:
nit: to decouple the test scenarios a new `client` mock and `provider` should be created for each iteration of this loop